dialer.go

```diff
@@ -0,0 +1,258 @@
+package relayproxy
+
+import (
+	"context"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	relaygrpc "github.com/bloXroute-Labs/relay-grpc"
+	"github.com/bloXroute-Labs/relayproxy/common"
+	"github.com/rs/zerolog"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/keepalive"
+)
+
+const (
+	windowSize = 1024 * 1024 * 10 // 3 MB
+	bufferSize = 1024 * 1024 * 8  // 8 MB
+
+)
+
+type Dialer struct {
+	DialerClients     *DialerClients
+	DialerConnections *DialerConnections
+	DialURL           *DialURL
+	IPConverter       map[string]common.URLOpts
+}
+
+type DialerClients struct {
+	mu                    sync.RWMutex
+	clients               []*common.Client
+	streamingClients      []*common.Client
+	streamingBlockClients []*common.Client
+	registrationClients   []*common.Client
+}
+
+func (d *DialerClients) CopyClients() []*common.Client {
+	d.mu.RLock()
+	defer d.mu.RUnlock()
+
+	copiedClients := make([]*common.Client, len(d.clients))
+	copy(copiedClients, d.clients)
+
+	return copiedClients
+}
+
+type DialerConnections struct {
+	Conns               []*grpc.ClientConn
+	StreamingConns      []*grpc.ClientConn
+	RegConns            []*grpc.ClientConn
+	StreamingBlockConns []*grpc.ClientConn
+}
+
+type DialURL struct {
+	RelayURL          string
+	StreamingURL      string
+	RegistrationURL   string
+	StreamingBlockURL string
+}
+
+func NewDialer(dialerOpts ...DialerOption) *Dialer {
+	dialer := &Dialer{
+		DialerClients: &DialerClients{
+			clients:               make([]*common.Client, 0),
+			streamingClients:      make([]*common.Client, 0),
+			streamingBlockClients: make([]*common.Client, 0),
+			registrationClients:   make([]*common.Client, 0),
+		},
+		DialerConnections: &DialerConnections{
+			Conns:               make([]*grpc.ClientConn, 0),
+			StreamingConns:      make([]*grpc.ClientConn, 0),
+			RegConns:            make([]*grpc.ClientConn, 0),
+			StreamingBlockConns: make([]*grpc.ClientConn, 0),
+		},
+		DialURL: &DialURL{},
+	}
+
+	for _, option := range dialerOpts {
+		option(dialer)
+	}
+
+	return dialer
+}
+
+func (d *Dialer) SetDialer(l zerolog.Logger) {
```

Comment:
the settings for streaming blocks is supposed to be different from the setting for the other streams